src/golang/lib/workflow/dag/workflow_dag.go

```diff
@@ -164,14 +152,41 @@ func NewWorkflowDag(
 	dbArtifacts := dbWorkflowDag.Artifacts
 	dbOperators := dbWorkflowDag.Operators
 
+	artifactIDToInputOpID := make(map[uuid.UUID]uuid.UUID, len(dbArtifacts))
+	opIDToMetadataPath := make(map[uuid.UUID]string, len(dbOperators))
+	opIDsByInputArtifact := make(map[uuid.UUID][]uuid.UUID, len(dbArtifacts))
+	for _, dbOperator := range dbOperators {
+		for _, outputArtifactID := range dbOperator.Outputs {
+			artifactIDToInputOpID[outputArtifactID] = dbOperator.Id
+		}
+		opIDToMetadataPath[dbOperator.Id] = utils.InitializePath(opExecMode == operator.Preview)
+
+		for _, inputArtifactID := range dbOperator.Inputs {
+			opIDs, ok := opIDsByInputArtifact[inputArtifactID]
+			if !ok {
+				opIDs = make([]uuid.UUID, 0, 1)
+			}
+			opIDsByInputArtifact[inputArtifactID] = append(opIDs, dbOperator.Id)
+		}
+	}
+
```

Comment:
The `workflow_dag` constructor is getting a little unwieldy. I think we need to do a little follow-up refactor to think through the minimum set of maps that we need, like we do in the SDK's dag representation.